File: data-algo/fetch-lc/utils/csv_analyze.js
import fs from 'fs'

import { parse } from 'csv-parse'
import { getRelativePath } from '../getPath'

const escapeCsvCell = (s) => {
    const ESCAPE_CHAR = '""'
    const hasCharNeedEscape = String(s).match(/,|"/)
    if (!hasCharNeedEscape) return s

    return `"${s.replace(/"/g, ESCAPE_CHAR)}"`
}

export const save_to_file = (filename, ext, data) => {
    if (ext === 'json') data = JSON.stringify(data, null, 2)

    const filePath = getRelativePath(`${filename}.${ext}`)
    fs.writeFileSync(filePath, data, 'utf8')

    console.log(`✅ Saved ${filename} to ${filePath}`)
}

export function save_csv(arr, filename) {
    // arr: string[][], row: string[]
    const content = arr
        .map((row) => row.map(escapeCsvCell).join(','))
        .join('\n')

    save_to_file(filename, 'csv', content)
}

export const read_stream_csv = (filename) =>
    new Promise((resolve, reject) => {
        const headers = {} // Record<string, number>
        const rows = []

        let isHeaderRead = false

        const processStream = (row) => {
            if (!isHeaderRead) {
                row.forEach((col, i) => (headers[col] = i))
                isHeaderRead = true
            } else {
                rows.push(row)
            }
        }

        fs.createReadStream(filename)
            .pipe(parse({ delimiter: ',' }))
            .on('data', processStream)
            .on('end', () => resolve({ rows, headers }))
            .on('error', reject)
    })

// =======================================
// ANALYSIS UTILS
export const initDiff = () => {
    const DIFFS = ['Easy', 'Medium', 'Hard']
    const diff_counts = {}

    for (const diff of DIFFS) {
        diff_counts[diff] = {
            count: 0,
            day: Array(7).fill(0)
        }
    }
    return diff_counts
}

export const round = (num) => {
    return parseFloat(num).toLocaleString('en-US', {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2,
        roundingMode: 'trunc'
    })
}

// =======================================
// SMALL UTILS
export const compare = (a, b) => {
    if (a > b) return 1
    if (a < b) return -1
    return 0
}

export function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms))
}

export const exit = (msg) => {
    console.error(msg)
    throw new Error(msg)
}

File: data-algo/fetch_py/utils_py/csv_analyze.py
import csv
import asyncio
import os
from pathlib import Path


def get_relative_path(filename):
    """Get relative path from current file's directory"""
    current_dir = Path(__file__).parent.parent
    return current_dir / filename


def save_file(arr, filename):
    """Save array data to CSV file"""
    file_path = get_relative_path(f'{filename}.csv')
    
    def escape_csv_cell(s):
        """Escape CSV special characters"""
        s_str = str(s) if s is not None else ''
        if ',' in s_str or '"' in s_str:
            return f'"{s_str.replace(chr(34), chr(34)+chr(34))}"'
        return s_str
    
    # row: list
    content = '\n'.join(
        [','.join([escape_csv_cell(cell) for cell in row]) for row in arr]
    )
    
    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)


def read_csv(filename):
    """Read CSV file and return headers and rows"""
    headers = {}
    rows = []
    
    with open(filename, 'r', encoding='utf-8') as f:
        reader = csv.reader(f)
        is_header_read = False
        
        for row in reader:
            if not is_header_read:
                for i, col in enumerate(row):
                    headers[col] = i
                is_header_read = True
            else:
                rows.append(row)
    
    return {'rows': rows, 'headers': headers}


# =======================================
# ANALYSIS UTILS
def init_diff():
    """Initialize difficulty counts structure"""
    DIFFS = ['Easy', 'Medium', 'Hard']
    diff_counts = {}
    
    for diff in DIFFS:
        diff_counts[diff] = {
            'count': 0,
            'day': [0] * 7
        }
    return diff_counts


def round_num(num):
    """Round number to 2 decimal places"""
    return f'{float(num):.2f}'


# =======================================
# SMALL UTILS
def compare(a, b):
    """Compare two values"""
    if a > b:
        return 1
    if a < b:
        return -1
    return 0


async def sleep_async(ms):
    """Async sleep in milliseconds"""
    await asyncio.sleep(ms / 1000)

